mysql/service_instance.go

```diff
@@ -0,0 +1,434 @@
+	CurrentOperation      map[string]string `json:"currentOperation"` //hope this works
+	TotalAvailablePatches string            `json:"totalAvailablePatches"`
+}
+
+func (c *ServiceInstanceClient) CreateServiceInstance(input *CreateServiceInstanceInput) (*ServiceInstance, error) {
+	var (
+		serviceInstance      *ServiceInstance
+		serviceInstanceError error
+	)
+
+	if c.PollInterval == 0 {
+		c.PollInterval = WaitForServiceInstanceReadyPollInterval
+	}
+	if c.Timeout == 0 {
+		c.Timeout = WaitForServiceInstanceReadyTimeout
+	}
+
+	// Since these CloudStorageUsername and CloudStoragePassword are sensitive we'll read them
+	// from the client if they haven't specified in the config.
+	if input.ServiceParameters.CloudStorageContainer != "" && input.ServiceParameters.CloudStorageUsername == "" && input.ServiceParameters.CloudStoragePassword == "" {
+		input.ServiceParameters.CloudStorageUsername = *c.ResourceClient.MySQLClient.client.UserName
+		input.ServiceParameters.CloudStoragePassword = *c.ResourceClient.MySQLClient.client.Password
+	}
+
+	for i := 0; i < *c.MySQLClient.client.MaxRetries; i++ {
+		serviceInstance, serviceInstanceError = c.startServiceInstance(input.ServiceParameters.ServiceName, input)
+		if serviceInstanceError == nil {
+			return serviceInstance, nil
+		}
+	}
+
+	return nil, serviceInstanceError
+}
+
+func (c *ServiceInstanceClient) startServiceInstance(name string, input *CreateServiceInstanceInput) (*ServiceInstance, error) {
+	if err := c.createServiceInstanceResource(*input, nil); err != nil {
+		return nil, err
+	}
+
+	// Call wait for instance ready now, as creating the instance is an eventually consistent operation
+	getInput := &GetServiceInstanceInput{
+		Name: name,
+	}
+
+	serviceInstance, serviceInstanceError := c.WaitForServiceInstanceRunning(getInput, c.PollInterval, c.Timeout)
+
+	if serviceInstanceError != nil {
+		c.client.DebugLogString(fmt.Sprintf("[Debug] : Create Failed %s", serviceInstanceError))
+		return nil, serviceInstanceError
+	}
+
+	return serviceInstance, nil
+}
+
+// GetServiceInstance retrieves the ServiceInstance with the given name.
+func (c *ServiceInstanceClient) GetServiceInstance(getInput *GetServiceInstanceInput) (*ServiceInstance, error) {
+	var serviceInstance ServiceInstance
+	if err := c.getServiceInstanceResource(getInput.Name, &serviceInstance); err != nil {
+		return nil, err
+	}
+
+	return &serviceInstance, nil
+}
+
+// WaitForInstanceReady waits for an instance to be completely initialized and available.
+func (c *ServiceInstanceClient) WaitForServiceInstanceRunning(input *GetServiceInstanceInput, pollingInterval time.Duration, timeoutSeconds time.Duration) (*ServiceInstance, error) {
+	var info *ServiceInstance
+	var getErr error
+
+	err := c.client.WaitFor("service instance to be ready", pollingInterval, timeoutSeconds, func() (bool, error) {
+		info, getErr = c.GetServiceInstance(input)
+		if getErr != nil {
+			return false, getErr
+		}
+		switch s := info.Status; s {
```

Comment:
Let's add logs under each case so we know what the status is whenever we poll. This is especially useful under the default case when we don't know the specific status is.

Reply:
Done